File: src/engine/chain_store.rs
use serde::{Deserialize, Serialize};

use std::collections::{HashMap};

use crate::engine::{chain::Chain, embed_chains::get_chains, event::Event, effect::Effect};

use super::utils::prefix_all_keys;

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct ChainStore {
    chains: HashMap<String, Chain>,
    events: HashMap<String, Event>,
    effects: HashMap<String, Effect>,
}

impl ChainStore {
    pub fn new(chain_files_override_paths: Vec<String>) -> ChainStore {
        let mut chains: HashMap<String, Chain> = Default::default();
        let mut events: HashMap<String, Event> = Default::default();
        let mut effects: HashMap<String, Effect> = Default::default();

        for mut json_chain in get_chains(chain_files_override_paths) {
            chains.insert(json_chain.title.clone(), json_chain.to_chain());

            let mut chain_events = prefix_all_keys(&mut json_chain.events, &(json_chain.title.clone() + "/"));

            for (_, event) in chain_events.iter_mut() {
                event.set_namespace(json_chain.title.clone())
            }

            events.extend(chain_events);
            effects.extend(prefix_all_keys(&mut json_chain.effects, &(json_chain.title.clone() + "/")));
        }

        return ChainStore {
            chains: chains,
            effects: effects,
            events: events,
        };
    }

    pub fn get_event(self, event_name: String) -> Event {
        return self.events.get(&event_name)
            .expect(("unknown event required: ".to_owned() + &event_name).as_str())
            .clone();
    }

    pub fn get_effect(self, effect_name: String) -> Effect {
        return self.effects.get(&effect_name)
            .expect(("unknown effect required: ".to_owned() + &effect_name).as_str())
            .clone();
    }

    pub fn get_chain(self, chain_name: String) -> Chain {
        return self.chains.get(&chain_name)
            .expect(("unknown chain required: ".to_owned() + &chain_name).as_str())
            .clone();
    }

    pub fn get_chains(&self) -> &HashMap<String, Chain> {
        &self.chains
    }
}

pub fn extract_chain_title(namespaced_id: &str) -> &str {
    namespaced_id.splitn(2, '/').next().unwrap_or("")
}

#[cfg(test)]
mod tests {
    use crate::engine::chain_store::extract_chain_title;

    #[test]
    fn test_extract_chain_title() {
        let input = "earthquake/report_about_destroyed_shelters";
        let result = extract_chain_title(input);
        assert_eq!(result, "earthquake");
    }
}

File: src/lib/engine/chain_store.ts
import {
	type Chain,
	type ChainEvent,
	type Effect,
	isStateVariable,
	parseStateVariable,
	type StateCondition,
	type StateVariable,
} from './model';
import {
	addNamespaceToIdentifier,
	addNamespaceToKeys,
	setNamespaceInEvent,
} from './namespace';
import { validate as validateJsonSchema } from 'jsonschema';

export class ChainStore {
	readonly chains: { [key: string]: Chain } = {};
	events: { [key: string]: ChainEvent } = {};
	effects: { [key: string]: Effect } = {};

	public getChain(title: string): Chain {
		return this.chains[title];
	}

	public getChains(): Array<Chain> {
		return Object.values(this.chains);
	}

	public getEvent(eventName: string): ChainEvent {
		return this.events[eventName];
	}

	public getEffect(effectName: string): Effect {
		return this.effects[effectName];
	}

	constructor(options?: ConstructorOptions) {
		const chainFiles = getChainsFiles(options);
		for (const chainFile in chainFiles) {
			const jsonChain = { ...chainFiles[chainFile] } as JSONChain;
			// eslint-disable-next-line @typescript-eslint/no-explicit-any
			delete (jsonChain as any).default;

			const validationRes = validateJsonSchema(jsonChain, getJsonSchema());
			if (!validationRes.valid)
				throw new Error(
					`invalid chain: ${chainFile}, json schema error:${JSON.stringify(validationRes.errors)}`,
				);

			this.chains[jsonChain.title] = {
				...jsonChain,
				cooldown: jsonChain.cooldown || 0,
				autoSelect: jsonChain.autoSelect || false,
				usedVariables: getUsedVariablesIn(jsonChain),
			};

			for (const eventId in jsonChain.events) {
				const event = jsonChain.events[eventId];
				setNamespaceInEvent(event, jsonChain.title);
				this.events[addNamespaceToIdentifier(eventId, jsonChain.title)] = event;
			}

			if (jsonChain.effects) {
				addNamespaceToKeys(jsonChain.effects, jsonChain.title);
				this.effects = {
					...this.effects,
					...jsonChain.effects,
				};
			}
		}
	}
}

function getJsonSchema() {
	const foundFiles = Object.values(
		import.meta.glob(['/chains/schema.json'], { eager: true }),
	);
	if (foundFiles.length === 0) {
		throw new Error('json schema missing');
	} else if (foundFiles.length > 1) {
		throw new Error('there was multiple json schema');
	}
	return foundFiles[0];
}

function getChainsFiles(options?: ConstructorOptions): Record<string, unknown> {
	if (options && options.overrideInputChains) {
		return options.overrideInputChains;
	}
	if (import.meta.env.VITE_ENABLE_TEST_ENV === 'true') {
		const urlParams = new URLSearchParams(window.location.search);
		const overrideInputChainFilenames = urlParams
			.getAll('overrideInputChainFilenames')
			.map((name) => `/test/chains/${name}.json`);
		const files: Record<string, unknown> = import.meta.glob(
			['/test/chains/*.json', '!**/schema.json'],
			{ eager: true },
		);

		return Object.fromEntries(
			Object.entries(files).filter(
				([fileName]) => overrideInputChainFilenames.indexOf(fileName) >= 0,
			),
		);
	}

	return import.meta.glob(['/chains/*.json', '!**/schema.json'], {
		eager: true,
	});
}

export type JSONChain = {
	title: string;
	cooldown?: number;
	trigger?: StateCondition;
	autoSelect?: boolean;
	events: {
		start: ChainEvent;
		[key: string]: ChainEvent;
	};
	effects?: {
		[key: string]: Effect;
	};
};

type ConstructorOptions = {
	overrideInputChains?: Record<string, unknown>;
};

export function getUsedVariablesIn(jsonChain: JSONChain): StateVariable[] {
	const keyToFind = 'target';
	const seen = new Set<StateVariable>();

	function recurse(jsonTree: unknown) {
		if (Array.isArray(jsonTree)) {
			for (const item of jsonTree) recurse(item);
		} else if (jsonTree && typeof jsonTree === 'object') {
			for (const [key, val] of Object.entries(jsonTree as Record<string, unknown>)) {
				if (
					key === keyToFind &&
					typeof val === 'string' &&
					isStateVariable(val)
				) {
					seen.add(parseStateVariable(String(val)));
				}
				recurse(val);
			}
		}
	}

	recurse(jsonChain);
	return Array.from(seen);
}
